app/assets/javascripts/govuk_publishing_components/analytics/auto-scroll-tracker.js

```diff
@@ -0,0 +1,195 @@
+    clearTimeout(this.resizeTimeout)
+    this.resizeTimeout = setTimeout(function () {
+      this.getWindowDetails()
+      this.trackVisibleNodes()
+    }.bind(this), this.config.resizeTimeoutDelay)
+  }
+
+  AutoScrollTracker.prototype.getWindowDetails = function () {
+    this.pageHeight = document.querySelector('body').clientHeight
+    this.windowHeight = window.innerHeight
+    this.trackedNodes = this.track.getTrackingNodes(this.trackedNodes)
+  }
+
+  AutoScrollTracker.prototype.trackVisibleNodes = function () {
+    for (var i = 0; i < this.trackedNodes.length; i++) {
+      var node = this.trackedNodes[i]
+      if (this.isVisible(node.top, node.bottom) && !node.alreadySeen) {
+        node.alreadySeen = true
+        // we store whether a heading has been tracked or not on the heading
+        // because if headings appear/disapper (e.g. inside an accordion)
+        // the order changes, so we can't refer to the previous trackedNodes
+        // as we do with percentages
+        if (node.node) {
+          node.node.setAttribute('data-autoscrolltracker-already-seen', true)
+        }
+
+        var action = node.eventData.action
+        var label = node.eventData.label
+
+        if (window.GOVUK.analytics && window.GOVUK.analytics.trackEvent) {
+          window.GOVUK.analytics.trackEvent('ScrollTo', action, { label: label, nonInteraction: true })
+        }
+      }
+    }
+  }
+
+  AutoScrollTracker.prototype.isVisible = function (top, bottom) {
+    var scroll = window.scrollY || document.documentElement.scrollTop // IE fallback
+    return scroll <= top && (scroll + this.windowHeight) >= bottom
+  }
+
+  AutoScrollTracker.Heading = function (config) {
+    this.config = config
+  }
+
+  AutoScrollTracker.Heading.prototype.getTrackingNodes = function () {
+    var headingsDetails = []
+    var headingsFound = this.findAllowedHeadings()
+
+    for (var i = 0; i < headingsFound.length; i++) {
+      var heading = headingsFound[i]
+      // only track headings that are visible i.e. not inside display: none
+      if (this.visible(heading)) {
+        var pos = heading.getBoundingClientRect()
+        headingsDetails.push({
+          node: heading,
+          alreadySeen: heading.getAttribute('data-autoscrolltracker-already-seen'),
+          top: pos.top + document.documentElement.scrollTop,
+          bottom: pos.bottom + document.documentElement.scrollTop,
+          eventData: { action: 'Heading', label: heading.textContent.replace(/\s+/g, ' ').trim() }
+        })
+      }
+    }
+    return headingsDetails
+  }
+
+  // check heading is inside allowed elements, generally ignores everything outside of page content
+  AutoScrollTracker.Heading.prototype.findAllowedHeadings = function () {
+    var headings = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6']
+    var headingsFound = []
+
+    for (var h = 0; h < this.config.allowHeadingsInside.length; h++) {
+      var inside = document.querySelector(this.config.allowHeadingsInside[h])
```

Comment:
I think you should adjust this to cater for a selector within `this.config.allowHeadingsInside` not existing - just incase we either have any rogue documents without a main or we expand that selector.

You may want to do it based on:

```
var insideElements = document.querySelectorAll(this.config.allowHeadingsInside)
for (var h = 0; h < insideElements.length; h++) {
  var found = insideElements[h].querySelectorAll(headings)
  // ...
}
```

Reply:
Good shout, have adjusted. It's a bit of a nest of loops, but it feels safer.